proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+### Why not attribute per type parameter
+Rather than applying `[SupportsOnlyNonByRefLike]` at the member level, it could be applied more granularly at a type parameter level. Consider for example: 
+
+```c#
+struct S<T, U> 
+{
+    where T : ref struct
+    where U : ref struct 
+
+    [SupportsOnlyNonByRefLike(nameof(T))]
+    void M(T[] array)
+    {
+
+    }
+}
+```
+
+In this example the method `M` is only invalid when `T` is a `ref struct`. The type of `U` is immaterial to the legality of the method as it's not used in a manner incompatible with `ref struct`. In this case it would be desirable to have a more granular opt out. 
+
+At this time though there are no practical examples of this pattern. Adding this granularity will add cost to the runtime, as well as to the language as it's hard to reference type parameters in attributes. Going for simplicity at the moment but can reconsider in future versions if evidence for the scenario appears.
+
+### Constraints vs. anti constraints
+The syntax `where T : ref struct` appears in the space C# refers to as constraints. That is the syntax is meant to constraint or limit the set of types that are legal for the generic type parameter.
+
+This feature though expands the set of types available. The set of types applicable to `T where T : ref struct` is larger than simply `T`. This means the feature is an anti-constraint yet it occupies a constraint location. 
+
+It is reasonable to consider a new syntax here that cleanly differentiates our anti-constraints from actual constraints. At the time of this writing though there are only two proposed anti-constraints and one is lacking supporting evidence. This makes it questionable if it meets the bar for a new syntax or not.
+
+If a new syntax is desired to separate anti-constraints from constraints then `allow` is one to consider: 
+
+```c#
+readonly ref struct Span<T> 
+    allow T : ref struct
+{
+
+}
+```
+
+This makes the support for `[SupportsOnlyNonByRefLike]` cleaner as the implementation is to simply ignore the set of `allow` items. It may even lead to a better name such as `[RemoveAllows]`. 
+
+The only downside is that it means type parameter declarations get more complex because there are now both `where` and `allow` lists on them.
+
+```c#
+void Write<T>(T value)
+    where T : ISpanFormattable
+    allow T : ref struct
+{
+
+}
+```
+
+At the same time it's possible this will read clearer to developers. It also opens the door for future allows such as pointers. 
```

Comment:
```suggestion
At the same time it's possible this will read clearer to developers. It also opens the door for future `allow`s such as pointers. 
```